dmenu.1
.TH DMENU 1 dmenu\-VERSION
.SH NAME
dmenu \- dynamic menu
.SH SYNOPSIS
.B dmenu
.RB [ \-b ]
.RB [ \-f ]
.RB [ \-i ]
.RB [ \-pe ]
.RB [ \-hf
.IR history-file ]
.RB [ \-l
.IR lines ]
.RB [ \-p
.IR prompt ]
.RB [ \-fn
.IR font ]
.RB [ \-nb
.IR color ]
.RB [ \-nf
.IR color ]
.RB [ \-sb
.IR color ]
.RB [ \-sf
.IR color ]
.RB [ \-v ]
.P
.BR dmenu_run " ..."
.SH DESCRIPTION
.B dmenu
is a dynamic menu for X, which reads a list of newline\-separated items from
stdin.  When the user selects an item and presses Return, their choice is printed
to stdout and dmenu terminates.  Entering text will narrow the items to those
matching the tokens in the input.
.SH OPTIONS
.TP
.B \-b
dmenu appears at the bottom of the screen.
.TP
.B \-f
dmenu grabs the keyboard before reading stdin.  This is faster, but will lock up
X until stdin reaches end\-of\-file.
.TP
.B \-i
dmenu matches menu items case insensitively.
.TP
.B \-pe
read user's $PATH for executables instead of stdin
.TP
.BI \-hf " history-file"
specify history file, which will sort results by mostly used on top
.TP
.BI \-l " lines"
dmenu lists items vertically, with the given number of lines.
.TP
.BI \-p " prompt"
defines the prompt to be displayed to the left of the input field.
.TP
.BI \-fn " font"
defines the font or font set used. eg. "fixed" or "Monospace-12:normal" (an xft font)
.TP
.BI \-nb " color"
defines the normal background color.
.IR #RGB ,
.IR #RRGGBB ,
and X color names are supported.
.TP
.BI \-nf " color"
defines the normal foreground color.
.TP
.BI \-sb " color"
defines the selected background color.
.TP
.BI \-sf " color"
defines the selected foreground color.
.TP
.B \-v
prints version information to stdout, then exits.
.SH USAGE
dmenu is completely controlled by the keyboard.  Items are selected using the
arrow keys, page up, page down, home, and end.
.TP
.B Tab
Copy the selected item to the input field.
.TP
.B Return
Confirm selection.  Prints the selected item to stdout and exits, returning
success.
.TP
.B Shift\-Return
Confirm input.  Prints the input text to stdout and exits, returning success.
.TP
.B Escape
Exit without selecting an item, returning failure.
.TP
C\-a
Home
.TP
C\-b
Left
.TP
C\-c
Escape
.TP
C\-d
Delete
.TP
C\-e
End
.TP
C\-f
Right
.TP
C\-g
Escape
.TP
C\-h
Backspace
.TP
C\-i
Tab
.TP
C\-j
Return
.TP
C\-J
Shift-Return
.TP
C\-k
Delete line right
.TP
C\-m
Return
.TP
C\-n
Down
.TP
C\-p
Up
.TP
C\-u
Delete line left
.TP
C\-w
Delete word left
.TP
C\-y
Paste from primary X selection
.TP
C\-Y
Paste from X clipboard
.TP
M\-g
Home
.TP
M\-G
End
.TP
M\-h
Up
.TP
M\-j
Page down
.TP
M\-k
Page up
.TP
M\-l
Down
.SH SEE ALSO
.IR dwm (1),
.IR stest (1)


dmenu.c
/* See LICENSE file for copyright and license details. */
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <unistd.h>
#include <dirent.h>
#include <errno.h>
#include <X11/Xlib.h>
#include <X11/Xatom.h>
#include <X11/Xutil.h>
#include <sys/stat.h>
#include <sys/types.h>
#ifdef XINERAMA
#include <X11/extensions/Xinerama.h>
#endif
#include "draw.h"

#define INTERSECT(x,y,w,h,r)  (MAX(0, MIN((x)+(w),(r).x_org+(r).width)  - MAX((x),(r).x_org)) \
                             * MAX(0, MIN((y)+(h),(r).y_org+(r).height) - MAX((y),(r).y_org)))
#define MIN(a,b)              ((a) < (b) ? (a) : (b))
#define MAX(a,b)              ((a) > (b) ? (a) : (b))
#define DEFFONT "fixed" /* xft example: "Monospace-11" */

typedef struct Item Item;
struct Item {
    char *text;
    int priority;
    Item *left, *right;
};

static void appenditem(Item *item, Item **list, Item **last);
static void calcoffsets(void);
static void cleanup(void);
static char *cistrstr(const char *s, const char *sub);
static void drawmenu(void);
static void grabkeyboard(void);
static void insert(const char *str, ssize_t n);
static void keypress(XKeyEvent *ev);
static void match(void);
static size_t nextrune(int inc);
static void paste(void);
static void readstdin(void);
static void run(void);
static void setup(void);
static void usage(void);
static void readPathExecutables(void);
static void readHistoryItems(const char *file);
static int sortByHist(const void *a, const void *b);
static void checkHistItemPriority(Item *item);
static void incHistoricalHits(char *text);

static char text[BUFSIZ] = "";
static int bh, mw, mh;
static int inputw, promptw;
static size_t cursor = 0;
static const char *font = NULL;
static const char *prompt = NULL;
static const char *normbgcolor = "#222222";
static const char *normfgcolor = "#bbbbbb";
static const char *selbgcolor  = "#005577";
static const char *selfgcolor  = "#eeeeee";
static unsigned int lines = 0;
static ColorSet *normcol;
static ColorSet *selcol;
static Atom clip, utf8;
static Bool topbar = True;
static Bool running = True;
static int ret = 0;
static DC *dc;
static Item *items = NULL;
static Item *matches, *matchend;
static Item *prev, *curr, *next, *sel;
static Window win;
static XIC xic;
static Bool pathexecutables = False;
static Item *hitems; /* history items */
static const char *history_file = NULL;
static int num_items = 0;

static int (*fstrncmp)(const char *, const char *, size_t) = strncmp;
static char *(*fstrstr)(const char *, const char *) = strstr;

int
main(int argc, char *argv[]) {
    Bool fast = False;
    int i;

    for(i = 1; i < argc; i++)
        /* these options take no arguments */
        if(!strcmp(argv[i], "-v")) {      /* prints version information */
            puts("dmenu-"VERSION", © 2006-2012 dmenu engineers, see LICENSE for details");
            exit(EXIT_SUCCESS);
        }
        else if(!strcmp(argv[i], "-b"))   /* appears at the bottom of the screen */
            topbar = False;
        else if(!strcmp(argv[i], "-pe"))  /* scan PATH executables */
            pathexecutables = True;
        else if(!strcmp(argv[i], "-f"))   /* grabs keyboard before reading stdin */
            fast = True;
        else if(!strcmp(argv[i], "-i")) { /* case-insensitive item matching */
            fstrncmp = strncasecmp;
            fstrstr = cistrstr;
        }
        else if(i+1 == argc)
            usage();
        /* these options take one argument */
        else if(!strcmp(argv[i], "-l"))   /* number of lines in vertical list */
            lines = atoi(argv[++i]);
        else if(!strcmp(argv[i], "-p"))   /* adds prompt to left of input field */
            prompt = argv[++i];
        else if(!strcmp(argv[i], "-fn"))  /* font or font set */
            font = argv[++i];
        else if(!strcmp(argv[i], "-nb"))  /* normal background color */
            normbgcolor = argv[++i];
        else if(!strcmp(argv[i], "-nf"))  /* normal foreground color */
            normfgcolor = argv[++i];
        else if(!strcmp(argv[i], "-sb"))  /* selected background color */
            selbgcolor = argv[++i];
        else if(!strcmp(argv[i], "-sf"))  /* selected foreground color */
            selfgcolor = argv[++i];
        else if(!strcmp(argv[i], "-hf"))  /* priority history file */
            history_file = argv[++i];
        else
            usage();

    dc = initdc();
    initfont(dc, font ? font : DEFFONT);
    normcol = initcolor(dc, normfgcolor, normbgcolor);
    selcol = initcolor(dc, selfgcolor, selbgcolor);

    /* read history file if set */
    if (history_file) {
        readHistoryItems(history_file);
    }

    if(fast) {
        grabkeyboard();
        if (pathexecutables) {
            readPathExecutables();
        } else {
            readstdin();
        }
    } else {
        if (pathexecutables) {
            readPathExecutables();
        } else {
            readstdin();
        }
        grabkeyboard();
    }
    /* sort by historical priority */
    if (items && history_file) {
        qsort(items, num_items, sizeof(struct Item), sortByHist);
    }
    setup();
    run();

    cleanup();
    return ret;
}

void
appenditem(Item *item, Item **list, Item **last) {
    if(*last)
        (*last)->right = item;
    else
        *list = item;

    item->left = *last;
    item->right = NULL;
    *last = item;
}

void
calcoffsets(void) {
    int i, n;

    if(lines > 0)
        n = lines * bh;
    else
        n = mw - (promptw + inputw + textw(dc, "<") + textw(dc, ">"));
    /* calculate which items will begin the next page and previous page */
    for(i = 0, next = curr; next; next = next->right)
        if((i += (lines > 0) ? bh : MIN(textw(dc, next->text), n)) > n)
            break;
    for(i = 0, prev = curr; prev && prev->left; prev = prev->left)
        if((i += (lines > 0) ? bh : MIN(textw(dc, prev->left->text), n)) > n)
            break;
}

char *
cistrstr(const char *s, const char *sub) {
    size_t len;

    for(len = strlen(sub); *s; s++)
        if(!strncasecmp(s, sub, len))
            return (char *)s;
    return NULL;
}

void
cleanup(void) {
    freecol(dc, normcol);
    freecol(dc, selcol);
    XDestroyWindow(dc->dpy, win);
    XUngrabKeyboard(dc->dpy, CurrentTime);
    freedc(dc);
}

void
drawmenu(void) {
    int curpos;
    Item *item;

    dc->x = 0;
    dc->y = 0;
    dc->h = bh;
    drawrect(dc, 0, 0, mw, mh, True, normcol->BG);

    if(prompt && *prompt) {
        dc->w = promptw;
        drawtext(dc, prompt, selcol);
        dc->x = dc->w;
    }
    /* draw input field */
    dc->w = (lines > 0 || !matches) ? mw - dc->x : inputw;
    drawtext(dc, text, normcol);
    if((curpos = textnw(dc, text, cursor) + dc->h/2 - 2) < dc->w)
        drawrect(dc, curpos, 2, 1, dc->h - 4, True, normcol->FG);

    if(lines > 0) {
        /* draw vertical list */
        dc->w = mw - dc->x;
        for(item = curr; item != next; item = item->right) {
            dc->y += dc->h;
            drawtext(dc, item->text, (item == sel) ? selcol : normcol);
        }
    }
    else if(matches) {
        /* draw horizontal list */
        dc->x += inputw;
        dc->w = textw(dc, "<");
        if(curr->left)
            drawtext(dc, "<", normcol);
        for(item = curr; item != next; item = item->right) {
            dc->x += dc->w;
            dc->w = MIN(textw(dc, item->text), mw - dc->x - textw(dc, ">"));
            drawtext(dc, item->text, (item == sel) ? selcol : normcol);
        }
        dc->w = textw(dc, ">");
        dc->x = mw - dc->w;
        if(next)
            drawtext(dc, ">", normcol);
    }
    mapdc(dc, win, mw, mh);
}

void
grabkeyboard(void) {
    int i;

    /* try to grab keyboard, we may have to wait for another process to ungrab */
    for(i = 0; i < 1000; i++) {
        if(XGrabKeyboard(dc->dpy, DefaultRootWindow(dc->dpy), True,
                         GrabModeAsync, GrabModeAsync, CurrentTime) == GrabSuccess)
            return;
        usleep(1000);
    }
    eprintf("cannot grab keyboard\n");
}

void
insert(const char *str, ssize_t n) {
    if(strlen(text) + n > sizeof text - 1)
        return;
    /* move existing text out of the way, insert new text, and update cursor */
    memmove(&text[cursor + n], &text[cursor], sizeof text - cursor - MAX(n, 0));
    if(n > 0)
        memcpy(&text[cursor], str, n);
    cursor += n;
    match();
}

void
keypress(XKeyEvent *ev) {
    char buf[32];
    int len;
    KeySym ksym = NoSymbol;
    Status status;

    len = XmbLookupString(xic, ev, buf, sizeof buf, &ksym, &status);
    if(status == XBufferOverflow)
        return;
    if(ev->state & ControlMask)
        switch(ksym) {
        case XK_a: ksym = XK_Home;      break;
        case XK_b: ksym = XK_Left;      break;
        case XK_c: ksym = XK_Escape;    break;
        case XK_d: ksym = XK_Delete;    break;
        case XK_e: ksym = XK_End;       break;
        case XK_f: ksym = XK_Right;     break;
        case XK_g: ksym = XK_Escape;    break;
        case XK_h: ksym = XK_BackSpace; break;
        case XK_i: ksym = XK_Tab;       break;
        case XK_j: /* fallthrough */
        case XK_J: ksym = XK_Return;    break;
        case XK_m: /* fallthrough */
        case XK_M: ksym = XK_Return;    break;
        case XK_n: ksym = XK_Down;      break;
        case XK_p: ksym = XK_Up;        break;

        case XK_k: /* delete right */
            text[cursor] = '\0';
            match();
            break;
        case XK_u: /* delete left */
            insert(NULL, 0 - cursor);
            break;
        case XK_w: /* delete word */
            while(cursor > 0 && text[nextrune(-1)] == ' ')
                insert(NULL, nextrune(-1) - cursor);
            while(cursor > 0 && text[nextrune(-1)] != ' ')
                insert(NULL, nextrune(-1) - cursor);
            break;
        case XK_y: /* paste selection */
            XConvertSelection(dc->dpy, (ev->state & ShiftMask) ? clip : XA_PRIMARY,
                              utf8, utf8, win, CurrentTime);
            return;
        default:
            return;
        }
    else if(ev->state & Mod1Mask)
        switch(ksym) {
        case XK_g: ksym = XK_Home;  break;
        case XK_G: ksym = XK_End;   break;
        case XK_h: ksym = XK_Up;    break;
        case XK_j: ksym = XK_Next;  break;
        case XK_k: ksym = XK_Prior; break;
        case XK_l: ksym = XK_Down;  break;
        default:
            return;
        }
    switch(ksym) {
    default:
        if(!iscntrl(*buf))
            insert(buf, len);
        break;
    case XK_Delete:
        if(text[cursor] == '\0')
            return;
        cursor = nextrune(+1);
        /* fallthrough */
    case XK_BackSpace:
        if(cursor == 0)
            return;
        insert(NULL, nextrune(-1) - cursor);
        break;
    case XK_End:
        if(text[cursor] != '\0') {
            cursor = strlen(text);
            break;
        }
        if(next) {
            /* jump to end of list and position items in reverse */
            curr = matchend;
            calcoffsets();
            curr = prev;
            calcoffsets();
            while(next && (curr = curr->right))
                calcoffsets();
        }
        sel = matchend;
        break;
    case XK_Escape:
        ret = EXIT_FAILURE;
        running = False;
    case XK_Home:
        if(sel == matches) {
            cursor = 0;
            break;
        }
        sel = curr = matches;
        calcoffsets();
        break;
    case XK_Left:
        if(cursor > 0 && (!sel || !sel->left || lines > 0)) {
            cursor = nextrune(-1);
            break;
        }
        if(lines > 0)
            return;
        /* fallthrough */
    case XK_Up:
        if(sel && sel->left && (sel = sel->left)->right == curr) {
            curr = prev;
            calcoffsets();
        }
        break;
    case XK_Next:
        if(!next)
            return;
        sel = curr = next;
        calcoffsets();
        break;
    case XK_Prior:
        if(!prev)
            return;
        sel = curr = prev;
        calcoffsets();
        break;
    case XK_Return:
    case XK_KP_Enter:
        puts((sel && !(ev->state & ShiftMask)) ? sel->text : text);
        if (history_file) {
            incHistoricalHits((sel && !(ev->state & ShiftMask)) ? sel->text : text);
        }
        ret = EXIT_SUCCESS;
        running = False;
    case XK_Right:
        if(text[cursor] != '\0') {
            cursor = nextrune(+1);
            break;
        }
        if(lines > 0)
            return;
        /* fallthrough */
    case XK_Down:
        if(sel && sel->right && (sel = sel->right) == next) {
            curr = next;
            calcoffsets();
        }
        break;
    case XK_Tab:
        if(!sel)
            return;
        strncpy(text, sel->text, sizeof text);
        cursor = strlen(text);
        match();
        break;
    }
    drawmenu();
}

void
match(void) {
    static char **tokv = NULL;
    static int tokn = 0;

    char buf[sizeof text], *s;
    int i, tokc = 0;
    size_t len;
    Item *item, *lprefix, *lsubstr, *prefixend, *substrend;

    strcpy(buf, text);
    /* separate input text into tokens to be matched individually */
    for(s = strtok(buf, " "); s; tokv[tokc-1] = s, s = strtok(NULL, " "))
        if(++tokc > tokn && !(tokv = realloc(tokv, ++tokn * sizeof *tokv)))
            eprintf("cannot realloc %u bytes\n", tokn * sizeof *tokv);
    len = tokc ? strlen(tokv[0]) : 0;

    matches = lprefix = lsubstr = matchend = prefixend = substrend = NULL;
    for(item = items; item && item->text; item++) {
        for(i = 0; i < tokc; i++)
            if(!fstrstr(item->text, tokv[i]))
                break;
        if(i != tokc) /* not all tokens match */
            continue;
        /* exact matches go first, then prefixes, then substrings */
        if(!tokc || !fstrncmp(tokv[0], item->text, len+1))
            appenditem(item, &matches, &matchend);
        else if(!fstrncmp(tokv[0], item->text, len))
            appenditem(item, &lprefix, &prefixend);
        else
            appenditem(item, &lsubstr, &substrend);
    }
    if(lprefix) {
        if(matches) {
            matchend->right = lprefix;
            lprefix->left = matchend;
        }
        else
            matches = lprefix;
        matchend = prefixend;
    }
    if(lsubstr) {
        if(matches) {
            matchend->right = lsubstr;
            lsubstr->left = matchend;
        }
        else
            matches = lsubstr;
        matchend = substrend;
    }
    curr = sel = matches;
    calcoffsets();
}

size_t
nextrune(int inc) {
    ssize_t n;

    /* return location of next utf8 rune in the given direction (+1 or -1) */
    for(n = cursor + inc; n + inc >= 0 && (text[n] & 0xc0) == 0x80; n += inc);
    return n;
}

void
paste(void) {
    char *p, *q;
    int di;
    unsigned long dl;
    Atom da;

    /* we have been given the current selection, now insert it into input */
    XGetWindowProperty(dc->dpy, win, utf8, 0, (sizeof text / 4) + 1, False,
                       utf8, &da, &di, &dl, &dl, (unsigned char **)&p);
    insert(p, (q = strchr(p, '\n')) ? q-p : (ssize_t)strlen(p));
    XFree(p);
    drawmenu();
}

void
readstdin(void) {
    char buf[sizeof text], *p, *maxstr = NULL;
    size_t i, max = 0, size = 0;

    /* read each line from stdin and add it to the item list */
    for(i = 0; fgets(buf, sizeof buf, stdin); i++) {
        if(i+1 >= size / sizeof *items)
            if(!(items = realloc(items, (size += BUFSIZ))))
                eprintf("cannot realloc %u bytes:", size);
        if((p = strchr(buf, '\n')))
            *p = '\0';
        if(!(items[i].text = strdup(buf)))
            eprintf("cannot strdup %u bytes:", strlen(buf)+1);
        if(strlen(items[i].text) > max)
            max = strlen(maxstr = items[i].text);
        items[i].priority = 0;
        checkHistItemPriority(&items[i]);

    }
    if(items)
        items[i].text = NULL;
    inputw = maxstr ? textw(dc, maxstr) : 0;
    lines = MIN(lines, i);
    num_items = i;
}

void
run(void) {
    XEvent ev;

    while(running && !XNextEvent(dc->dpy, &ev)) {
        if(XFilterEvent(&ev, win))
            continue;
        switch(ev.type) {
        case Expose:
            if(ev.xexpose.count == 0)
                mapdc(dc, win, mw, mh);
            break;
        case KeyPress:
            keypress(&ev.xkey);
            break;
        case SelectionNotify:
            if(ev.xselection.property == utf8)
                paste();
            break;
        case VisibilityNotify:
            if(ev.xvisibility.state != VisibilityUnobscured)
                XRaiseWindow(dc->dpy, win);
            break;
        }
    }
}

void
setup(void) {
    int x, y, screen = DefaultScreen(dc->dpy);
    Window root = RootWindow(dc->dpy, screen);
    XSetWindowAttributes swa;
    XIM xim;
#ifdef XINERAMA
    int n;
    XineramaScreenInfo *info;
#endif

    clip = XInternAtom(dc->dpy, "CLIPBOARD",   False);
    utf8 = XInternAtom(dc->dpy, "UTF8_STRING", False);

    /* calculate menu geometry */
    bh = dc->font.height + 2;
    lines = MAX(lines, 0);
    mh = (lines + 1) * bh;
#ifdef XINERAMA
    if((info = XineramaQueryScreens(dc->dpy, &n))) {
        int a, j, di, i = 0, area = 0;
        unsigned int du;
        Window w, pw, dw, *dws;
        XWindowAttributes wa;

        XGetInputFocus(dc->dpy, &w, &di);
        if(w != root && w != PointerRoot && w != None) {
            /* find top-level window containing current input focus */
            do {
                if(XQueryTree(dc->dpy, (pw = w), &dw, &w, &dws, &du) && dws)
                    XFree(dws);
            } while(w != root && w != pw);
            /* find xinerama screen with which the window intersects most */
            if(XGetWindowAttributes(dc->dpy, pw, &wa))
                for(j = 0; j < n; j++)
                    if((a = INTERSECT(wa.x, wa.y, wa.width, wa.height, info[j])) > area) {
                        area = a;
                        i = j;
                    }
        }
        /* no focused window is on screen, so use pointer location instead */
        if(!area && XQueryPointer(dc->dpy, root, &dw, &dw, &x, &y, &di, &di, &du))
            for(i = 0; i < n; i++)
                if(INTERSECT(x, y, 1, 1, info[i]))
                    break;

        x = info[i].x_org;
        y = info[i].y_org + (topbar ? 0 : info[i].height - mh);
        mw = info[i].width;
        XFree(info);
    }
    else
#endif
    {
        x = 0;
        y = topbar ? 0 : DisplayHeight(dc->dpy, screen) - mh;
        mw = DisplayWidth(dc->dpy, screen);
    }
    promptw = (prompt && *prompt) ? textw(dc, prompt) : 0;
    inputw = MIN(inputw, mw/3);
    match();

    /* create menu window */
    swa.override_redirect = True;
    swa.background_pixel = normcol->BG;
    swa.event_mask = ExposureMask | KeyPressMask | VisibilityChangeMask;
    win = XCreateWindow(dc->dpy, root, x, y, mw, mh, 0,
                        DefaultDepth(dc->dpy, screen), CopyFromParent,
                        DefaultVisual(dc->dpy, screen),
                        CWOverrideRedirect | CWBackPixel | CWEventMask, &swa);

    /* open input methods */
    xim = XOpenIM(dc->dpy, NULL, NULL, NULL);
    xic = XCreateIC(xim, XNInputStyle, XIMPreeditNothing | XIMStatusNothing,
                    XNClientWindow, win, XNFocusWindow, win, NULL);

    XMapRaised(dc->dpy, win);
    resizedc(dc, mw, mh);
    drawmenu();
}

void
usage(void) {
    fputs("usage: dmenu [-b] [-f] [-i] [-pe] [-l lines] [-p prompt] [-fn font]\n"
          "             [-nb color] [-nf color] [-sb color] [-sf color]\n"
          "             [-hf history file] [-v]\n", stderr);
    exit(EXIT_FAILURE);
}

void
readPathExecutables() {
    char *dirNamePtr, *path = getenv("PATH");
    size_t i = 0, size = 0, max = 0;
    DIR *dirp = NULL;
    struct dirent *dp;
    char fullpath[PATH_MAX], *end, *maxstr = NULL;

    /* actually all read executables can be cached,
     * but on my machine its not a problem at all */
    if (path == NULL) {
        perror("getenv");
        exit(EXIT_FAILURE);
    }

    /* split path into directory list */
    dirNamePtr = strtok(path, ":");
    while (dirNamePtr != NULL) {
        /* read directory */
        dirp = opendir(dirNamePtr);

        strcpy(fullpath, dirNamePtr);
        end = fullpath + strlen(fullpath);

        if (dirp == NULL) {
            if (errno != ENOENT) {
                perror("opendir");
            }
            continue;
        }
        dp = readdir(dirp);
        while (dp != NULL) {
            strcat(end, "/");
            strcpy(end + 1, dp->d_name);
            struct stat sb;
            if (!stat(fullpath, &sb) && S_ISREG(sb.st_mode) && sb.st_mode & 0111) {
                if (i + 1 >= size / sizeof *items) {
                    if (!(items = realloc(items, (size += BUFSIZ)))) {
                        eprintf("cannot realloc %u bytes:", (unsigned int)size);
                        exit(EXIT_FAILURE);
                    }
                }
                if (!(items[i].text = strdup(dp->d_name))) {
                    eprintf("cannot strdup %u bytes:", (unsigned int)strlen(dp->d_name) + 1);
                    exit(EXIT_FAILURE);
                }
                if (strlen(items[i].text) > max) {
                    max = strlen(maxstr = items[i].text);
                }
                items[i].priority = 0;
                checkHistItemPriority(&items[i]);
                i += 1;
            }
            dp = readdir(dirp);
        }
        closedir(dirp);
        dirNamePtr = strtok(NULL, ":");
    }

    if (items) {
        items[i].text = NULL;
    }
    inputw = maxstr ? textw(dc, maxstr) : 0;
    lines = MIN(lines, i);
    num_items = i;
}

void
readHistoryItems(const char *file) {
    FILE *in;
    char buf[BUFSIZ], *pch, *p;
    size_t i, size = 0;

    in = fopen(file, "r");
    if (!in) {
        return; /* no history file yet, will generate */
    }

    for (i = 0; fgets(buf, sizeof buf, in); i++) {
        if (i+1 >= size / sizeof *hitems) {
            if (!(hitems = realloc(hitems, (size += BUFSIZ)))) {
                eprintf("cannot realloc %u bytes:", (unsigned int)size);
                exit(EXIT_FAILURE);
            }
        }
        if ((p = strchr(buf, '\n'))) {
            *p = '\0';
        }
        pch = strtok(buf, "@");
        if (pch != NULL) {
            if (!(hitems[i].text = strdup(pch))) {
                eprintf("cannot strdup %u bytes:", (unsigned int)strlen(pch)+1);
                exit(EXIT_FAILURE);
            }
            hitems[i].priority = atoi(strtok(NULL, "@"));
        }
    }
    if (hitems) {
        hitems[i].text = NULL;
    }
}

void
checkHistItemPriority(Item *item) {
    Item *hitem;
    for (hitem = hitems; hitem && hitem->text; hitem++) {
        if (!strcmp(hitem->text, item->text)) {
            item->priority = hitem->priority;
            return;
        }
    }
}

int
sortByHist(const void *a, const void *b) {
    struct Item *ia = (struct Item *)a;
    struct Item *ib = (struct Item *)b;
    return ib->priority - ia->priority;
}

void
incHistoricalHits(char *text) {
    FILE *out;
    Item *hitem;
    int written = 0;

    out = fopen(history_file, "w");
    if (!out) {
        perror("fopen");
        exit(EXIT_FAILURE);
    }
    for (hitem = hitems; hitem && hitem->text; hitem++) {
        if (!strcmp(hitem->text, text)) {
            hitem->priority += 1;
            written = 1;
        }
        fprintf(out, "%s@%d\n", hitem->text, hitem->priority);
    }
    if (!written) {
        fprintf(out, "%s@%d\n", text, 1);
    }
    fclose(out);
}

